File: proj.android_studio/ee-x-admob/src/main/java/com/ee/admob/AdMobNativeAd.java
package com.ee.admob;

import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Point;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.RatingBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.ee.ads.AdViewHelper;
import com.ee.ads.IAdView;
import com.ee.ads.MessageHelper;
import com.ee.ads.ViewHelper;
import com.ee.core.IMessageBridge;
import com.ee.core.Logger;
import com.ee.core.internal.Thread;
import com.ee.core.internal.Utils;
import com.google.android.gms.ads.AdListener;
import com.google.android.gms.ads.AdLoader;
import com.google.android.gms.ads.AdRequest;
import com.google.android.gms.ads.VideoController;
import com.google.android.gms.ads.formats.MediaView;
import com.google.android.gms.ads.formats.NativeAd;
import com.google.android.gms.ads.formats.UnifiedNativeAd;
import com.google.android.gms.ads.formats.UnifiedNativeAdView;

import java.util.List;
import java.util.Map;

import static com.google.common.truth.Truth.assertThat;

/**
 * Created by Zinge on 10/16/17.
 */

class AdMobNativeAd extends AdListener implements IAdView {
    private static final Logger _logger = new Logger(AdMobNativeAd.class.getName());

    private static final String k__body = "body";
    private static final String k__call_to_action = "call_to_action";
    private static final String k__headline = "headline";
    private static final String k__icon = "icon";
    private static final String k__image = "image";
    private static final String k__media = "media";
    private static final String k__price = "price";
    private static final String k__star_rating = "star_rating";
    private static final String k__store = "store";

    private Context _context;
    private Activity _activity;
    private IMessageBridge _bridge;
    private String _adId;
    private String _layoutName;
    private Map<String, String> _identifiers;
    private MessageHelper _messageHelper;
    private AdViewHelper _helper;
    private ViewHelper _viewHelper;
    private boolean _isLoaded;
    private AdLoader _ad;
    private FrameLayout _view;

    AdMobNativeAd(@NonNull Context context,
                  @Nullable Activity activity,
                  @NonNull IMessageBridge bridge,
                  @NonNull String adId,
                  @NonNull String layoutName,
                  @NonNull Map<String, String> identifiers) {
        _logger.info("constructor: adId = %s", adId);
        Thread.checkMainThread();
        _context = context;
        _activity = activity;
        _bridge = bridge;
        _adId = adId;
        _layoutName = layoutName;
        _identifiers = identifiers;
        _messageHelper = new MessageHelper("AdMobNativeAd", adId);
        _helper = new AdViewHelper(_bridge, this, _messageHelper);

        createInternalAd();
        createView();
        registerHandlers();
    }

    void onCreate(@NonNull Activity activity) {
        _activity = activity;
        addToActivity(activity);
    }

    void onDestroy(@NonNull Activity activity) {
        assertThat(_activity).isEqualTo(activity);
        removeFromActivity(activity);
        _activity = null;
    }

    void destroy() {
        _logger.info("destroy: adId = " + _adId);
        Thread.checkMainThread();
        deregisterHandlers();
        destroyView();
        destroyInternalAd();
        _context = null;
        _bridge = null;
        _adId = null;
        _layoutName = null;
        _identifiers = null;
        _messageHelper = null;
        _helper = null;
    }

    private void registerHandlers() {
        _helper.registerHandlers();
    }

    private void deregisterHandlers() {
        _helper.deregisterHandlers();
    }

    private boolean createInternalAd() {
        Thread.checkMainThread();
        if (_ad != null) {
            return false;
        }
        _isLoaded = false;
        _ad = new AdLoader.Builder(_context, _adId)
            .forUnifiedNativeAd(nativeAppInstallAd -> {
                _logger.info("onAppInstallAdLoaded");
                int layoutId = _context.getResources().getIdentifier(_layoutName, "layout", _context.getPackageName());
                UnifiedNativeAdView adView = (UnifiedNativeAdView) LayoutInflater
                    .from(_context)
                    .inflate(layoutId, null, false);
                populateUnifiedNativeAdView(nativeAppInstallAd, adView);
                _view.removeAllViews();

                ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(
                    ViewGroup.LayoutParams.MATCH_PARENT,
                    ViewGroup.LayoutParams.MATCH_PARENT);
                _view.addView(adView, params);

                _isLoaded = true;
                _bridge.callCpp(_messageHelper.onLoaded());
            }).withAdListener(this).build();
        return true;
    }

    private boolean destroyInternalAd() {
        Thread.checkMainThread();
        if (_ad == null) {
            return false;
        }
        _isLoaded = false;
        _ad = null;
        return true;
    }

    private void createView() {
        Thread.checkMainThread();
        assertThat(_view).isNull();
        FrameLayout layout = new FrameLayout(_context);

        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT);
        params.gravity = Gravity.START | Gravity.TOP;
        layout.setLayoutParams(params);

        _view = layout;
        _viewHelper = new ViewHelper(_view);

        if (_activity != null) {
            addToActivity(_activity);
        }
    }

    private void destroyView() {
        Thread.checkMainThread();
        assertThat(_view).isNotNull();
        if (_activity != null) {
            removeFromActivity(_activity);
        }
        _viewHelper = null;
        _view = null;
    }

    private void addToActivity(@NonNull Activity activity) {
        FrameLayout rootView = Utils.getRootView(activity);
        rootView.addView(_view);
    }

    private void removeFromActivity(@NonNull Activity activity) {
        FrameLayout rootView = Utils.getRootView(activity);
        rootView.removeView(_view);
    }

    @Override
    public boolean isLoaded() {
        Thread.checkMainThread();
        assertThat(_ad).isNotNull();
        return _isLoaded;
    }

    @Override
    public void load() {
        Thread.checkMainThread();
        assertThat(_ad).isNotNull();
        _logger.info("load");
        AdRequest.Builder builder = new AdRequest.Builder();
        _ad.loadAd(builder.build());
    }

    @NonNull
    @Override
    public Point getPosition() {
        return _viewHelper.getPosition();
    }

    @Override
    public void setPosition(@NonNull Point position) {
        _viewHelper.setPosition(position);
    }

    @NonNull
    @Override
    public Point getSize() {
        return _viewHelper.getSize();
    }

    @Override
    public void setSize(@NonNull Point size) {
        _viewHelper.setSize(size);
    }

    @Override
    public boolean isVisible() {
        return _viewHelper.isVisible();
    }

    @Override
    public void setVisible(boolean visible) {
        _viewHelper.setVisible(visible);
    }

    private int getIdentifier(@NonNull String identifier) {
        if (!_identifiers.containsKey(identifier)) {
            return 0;
        }
        Resources resources = _context.getResources();
        return resources.getIdentifier(_identifiers.get(identifier), "id", _context.getPackageName());
    }

    private void populateUnifiedNativeAdView(UnifiedNativeAd nativeAppInstallAd,
                                             UnifiedNativeAdView adView) {
        // Get the video controller for the ad. One will always be provided, even if the ad doesn't
        // have a video asset.
        VideoController vc = nativeAppInstallAd.getVideoController();

        // Create a new VideoLifecycleCallbacks object and pass it to the VideoController. The
        // VideoController will call methods on this object when events occur in the video
        // lifecycle.
        vc.setVideoLifecycleCallbacks(new VideoController.VideoLifecycleCallbacks() {
            public void onVideoEnd() {
                // Publishers should allow native ads to complete video playback before refreshing
                // or replacing them with another ad in the same UI location.
                super.onVideoEnd();
            }
        });

        // Some assets are guaranteed to be in every NativeAppInstallAd.
        processView(adView, k__body, (ViewProcessor<TextView>) view -> {
            adView.setBodyView(view);
            view.setText(nativeAppInstallAd.getBody());
        });

        processView(adView, k__call_to_action, (ViewProcessor<Button>) view -> {
            adView.setCallToActionView(view);
            view.setText(nativeAppInstallAd.getCallToAction());
        });

        processView(adView, k__headline, (ViewProcessor<TextView>) view -> {
            adView.setHeadlineView(view);
            view.setText(nativeAppInstallAd.getHeadline());
        });

        processView(adView, k__icon, (ViewProcessor<ImageView>) view -> {
            adView.setIconView(view);
            if (nativeAppInstallAd.getIcon() != null) {
                view.setImageDrawable(nativeAppInstallAd.getIcon().getDrawable());
            }
        });

        // Apps can check the VideoController's hasVideoContent property to determine if the
        // NativeAppInstallAd has a video asset.
        if (vc.hasVideoContent()) {
            processView(adView, k__image, (ViewProcessor<ImageView>) view ->
                view.setVisibility(View.GONE));
            processView(adView, k__media, adView::setMediaView);
        } else {
            processView(adView, k__image, (ViewProcessor<ImageView>) view -> {
                List<NativeAd.Image> images = nativeAppInstallAd.getImages();
                view.setVisibility(View.GONE);
                for (NativeAd.Image image : images) {
                    if (image != null) {
                        view.setVisibility(View.VISIBLE);
                        view.setImageDrawable(image.getDrawable());
                        break;
                    }
                }
            });
            processView(adView, k__media, (ViewProcessor<MediaView>) view ->
                view.setVisibility(View.GONE));
        }

        // These assets aren't guaranteed to be in every NativeAppInstallAd, so it's important to
        // check before trying to display them.
        processView(adView, k__price, (ViewProcessor<TextView>) view -> {
            adView.setPriceView(view);
            CharSequence price = nativeAppInstallAd.getPrice();
            if (price != null) {
                view.setVisibility(View.VISIBLE);
                view.setText(price);
            } else {
                view.setVisibility(View.INVISIBLE);
            }
        });

        processView(adView, k__star_rating, (ViewProcessor<RatingBar>) view -> {
            adView.setStarRatingView(view);
            Double starRating = nativeAppInstallAd.getStarRating();
            if (starRating != null) {
                view.setRating(starRating.floatValue());
                view.setVisibility(View.VISIBLE);
            } else {
                view.setVisibility(View.INVISIBLE);
            }
        });

        processView(adView, k__store, (ViewProcessor<TextView>) view -> {
            adView.setStoreView(view);
            CharSequence store = nativeAppInstallAd.getStore();
            if (store != null) {
                view.setVisibility(View.VISIBLE);
                view.setText(store);
            } else {
                view.setVisibility(View.INVISIBLE);
            }
        });

        // Assign native ad object to the native view.
        adView.setNativeAd(nativeAppInstallAd);
    }

    private interface ViewProcessor<T extends View> {
        void process(T view);
    }

    @SuppressWarnings("unchecked")
    private <T extends View> void processView(@NonNull View view,
                                              @NonNull String key,
                                              @NonNull ViewProcessor<T> processor) {
        int id = getIdentifier(key);
        if (id == 0) {
            _logger.error("Can not find identifier for key: " + key);
            return;
        }
        View subView = view.findViewById(id);
        if (subView == null) {
            _logger.error("Can not find view for key: " + key);
            return;
        }
        processor.process((T) subView);
    }

    @Override
    public void onAdClosed() {
        _logger.info("onAdClosed");
        Thread.checkMainThread();
    }

    @Override
    public void onAdFailedToLoad(int errorCode) {
        _logger.info("onAdFailedToLoad: code = " + errorCode);
        Thread.checkMainThread();
        _bridge.callCpp(_messageHelper.onFailedToLoad(), String.valueOf(errorCode));
    }

    @Override
    public void onAdLeftApplication() {
        _logger.info("onAdLeftApplication");
        Thread.checkMainThread();
        _bridge.callCpp(_messageHelper.onClicked());
    }

    @Override
    public void onAdOpened() {
        _logger.info("onAdOpened");
        Thread.checkMainThread();
    }

    @Override
    public void onAdLoaded() {
        _logger.info("onAdLoaded");
        Thread.checkMainThread();
        // There is one important difference between the way AdListener objects work with native ads
        // and the way they work with banners and interstitials. Because the AdLoader has its own
        // format-specific listeners (i.e., UnifiedNativeAd.OnUnifiedNativeAdLoadedListener) to use
        // when an ad has loaded, the onAdLoaded() method from AdListener is not called when a
        // native ad loads successfully.
    }

    @Override
    public void onAdClicked() {
        _logger.info("onAdClicked");
        Thread.checkMainThread();
    }

    @Override
    public void onAdImpression() {
        _logger.info("onAdImpression");
        Thread.checkMainThread();
    }
}

File: proj.android_studio/ee-x-admob/src/main/java/com/ee/admob/AdMobNativeAd.kt
package com.ee.admob

import android.app.Activity
import android.content.Context
import android.graphics.Point
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.RatingBar
import android.widget.TextView
import androidx.annotation.AnyThread
import androidx.annotation.UiThread
import com.ee.ads.AdViewHelper
import com.ee.ads.IAdView
import com.ee.ads.MessageHelper
import com.ee.ads.ViewHelper
import com.ee.core.IMessageBridge
import com.ee.core.Logger
import com.ee.core.internal.Thread
import com.ee.core.internal.Utils
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.VideoController.VideoLifecycleCallbacks
import com.google.android.gms.ads.formats.MediaView
import com.google.android.gms.ads.formats.UnifiedNativeAd
import com.google.android.gms.ads.formats.UnifiedNativeAdView
import com.google.common.truth.Truth.assertThat
import kotlinx.serialization.ImplicitReflectionSerializer
import kotlinx.serialization.UnstableDefault
import java.util.concurrent.atomic.AtomicBoolean

private typealias ViewProcessor<T> = (view: T) -> Unit

/**
 * Created by Zinge on 10/16/17.
 */
@ImplicitReflectionSerializer
@UnstableDefault
internal class AdMobNativeAd(
    private val _bridge: IMessageBridge,
    private val _context: Context,
    private var _activity: Activity?,
    private val _adId: String,
    private val _layoutName: String,
    private val _identifiers: Map<String, String>)
    : AdListener()
    , IAdView {
    companion object {
        private val _logger = Logger(AdMobNativeAd::class.java.name)
        private const val k__body = "body"
        private const val k__call_to_action = "call_to_action"
        private const val k__headline = "headline"
        private const val k__icon = "icon"
        private const val k__image = "image"
        private const val k__media = "media"
        private const val k__price = "price"
        private const val k__star_rating = "star_rating"
        private const val k__store = "store"
    }

    private val _messageHelper = MessageHelper("AdMobNativeAd", _adId)
    private val _helper = AdViewHelper(_bridge, this, _messageHelper)
    private val _isLoaded = AtomicBoolean(false)
    private var _ad: AdLoader? = null
    private var _view: FrameLayout? = null
    private var _viewHelper: ViewHelper? = null

    init {
        _logger.info("constructor: adId = %s", _adId)
        registerHandlers()
        createInternalAd()
        createView()
    }

    fun onCreate(activity: Activity) {
        _activity = activity
        addToActivity()
    }

    fun onDestroy(activity: Activity) {
        assertThat(_activity).isEqualTo(activity)
        removeFromActivity()
        _activity = null
    }

    @AnyThread
    fun destroy() {
        _logger.info("${this::destroy}: adId = $_adId")
        deregisterHandlers()
        destroyView()
        destroyInternalAd()
    }

    @AnyThread
    private fun registerHandlers() {
        _helper.registerHandlers()
    }

    @AnyThread
    private fun deregisterHandlers() {
        _helper.deregisterHandlers()
    }

    @AnyThread
    private fun createInternalAd() {
        Thread.runOnMainThread(Runnable {
            if (_ad != null) {
                return@Runnable
            }
            _isLoaded.set(false)
            _ad = AdLoader.Builder(_context, _adId)
                .forUnifiedNativeAd { nativeAppInstallAd ->
                    _logger.info("onAppInstallAdLoaded")
                    val layoutId = _context.resources.getIdentifier(_layoutName, "layout", _context.packageName)
                    val adView = LayoutInflater
                        .from(_context)
                        .inflate(layoutId, null, false) as UnifiedNativeAdView
                    populateUnifiedNativeAdView(nativeAppInstallAd, adView)
                    _view?.removeAllViews()
                    val params = ViewGroup.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.MATCH_PARENT)
                    _view?.addView(adView, params)
                    _isLoaded.set(true)
                    _bridge.callCpp(_messageHelper.onLoaded)
                }.withAdListener(this).build()
        })
    }

    @AnyThread
    private fun destroyInternalAd() {
        Thread.runOnMainThread(Runnable {
            if (_ad == null) {
                return@Runnable
            }
            _isLoaded.set(false)
            _ad = null
        })
    }

    @AnyThread
    private fun createView() {
        Thread.runOnMainThread(Runnable {
            assertThat(_view).isNull()
            val view = FrameLayout(_context)
            val params = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT)
            params.gravity = Gravity.START or Gravity.TOP
            view.layoutParams = params
            _view = view
            _viewHelper = ViewHelper(view)
            addToActivity()
        })
    }

    @AnyThread
    private fun destroyView() {
        Thread.runOnMainThread(Runnable {
            removeFromActivity()
            _view = null
            _viewHelper = null
        })
    }

    @AnyThread
    private fun addToActivity() {
        Thread.runOnMainThread(Runnable {
            val activity = _activity ?: return@Runnable
            val rootView = Utils.getRootView(activity)
            rootView.addView(_view)
        })
    }

    @UiThread
    private fun removeFromActivity() {
        Thread.runOnMainThread(Runnable {
            val activity = _activity ?: return@Runnable
            val rootView = Utils.getRootView(activity)
            rootView.removeView(_view)
        })
    }

    override val isLoaded: Boolean
        @AnyThread get() = _isLoaded.get()

    @AnyThread
    override fun load() {
        Thread.runOnMainThread(Runnable {
            _logger.info("${this::load}")
            val ad = _ad ?: throw IllegalArgumentException("Ad is not initialized")
            ad.loadAd(AdRequest.Builder().build())
        })
    }

    override var position: Point
        @AnyThread get() = _viewHelper?.position ?: Point(0, 0)
        @AnyThread set(value) {
            _viewHelper?.position = value
        }

    override var size: Point
        @AnyThread get() = _viewHelper?.size ?: Point(0, 0)
        @AnyThread set(value) {
            _viewHelper?.size = value
        }

    override var isVisible: Boolean
        @AnyThread get() = _viewHelper?.isVisible ?: false
        @AnyThread set(value) {
            _viewHelper?.isVisible = value
        }

    private fun getIdentifier(identifier: String): Int {
        if (!_identifiers.containsKey(identifier)) {
            return 0
        }
        val resources = _context.resources
        return resources.getIdentifier(_identifiers[identifier], "id", _context.packageName)
    }

    private fun populateUnifiedNativeAdView(nativeAppInstallAd: UnifiedNativeAd,
                                            adView: UnifiedNativeAdView) {
        // Get the video controller for the ad. One will always be provided, even if the ad doesn't
        // have a video asset.
        val vc = nativeAppInstallAd.videoController

        // Create a new VideoLifecycleCallbacks object and pass it to the VideoController. The
        // VideoController will call methods on this object when events occur in the video
        // lifecycle.
        vc.videoLifecycleCallbacks = object : VideoLifecycleCallbacks() {
            override fun onVideoEnd() {
                // Publishers should allow native ads to complete video playback before refreshing
                // or replacing them with another ad in the same UI location.
                _logger.debug("${this::onVideoEnd}")
                super.onVideoEnd()
            }
        }

        // Some assets are guaranteed to be in every NativeAppInstallAd.
        processView(adView, k__body) { item: TextView ->
            adView.bodyView = item
            item.text = nativeAppInstallAd.body
        }
        processView(adView, k__call_to_action) { item: Button ->
            adView.callToActionView = item
            item.text = nativeAppInstallAd.callToAction
        }
        processView(adView, k__headline) { item: TextView ->
            adView.headlineView = item
            item.text = nativeAppInstallAd.headline
        }
        processView(adView, k__icon) { item: ImageView ->
            adView.iconView = item
            if (nativeAppInstallAd.icon != null) {
                item.setImageDrawable(nativeAppInstallAd.icon.drawable)
            }
        }

        // Apps can check the VideoController's hasVideoContent property to determine if the
        // NativeAppInstallAd has a video asset.
        if (vc.hasVideoContent()) {
            processView<ImageView>(adView, k__image) { item ->
                item.visibility = View.GONE
            }
            processView(adView, k__media) { item: MediaView ->
                adView.mediaView = item
            }
        } else {
            processView<ImageView>(adView, k__image) { view ->
                val images = nativeAppInstallAd.images
                view.visibility = View.GONE
                for (image in images) {
                    if (image != null) {
                        view.visibility = View.VISIBLE
                        view.setImageDrawable(image.drawable)
                        break
                    }
                }
            }
            processView<MediaView>(adView, k__media) { view ->
                view.visibility = View.GONE
            }
        }

        // These assets aren't guaranteed to be in every NativeAppInstallAd, so it's important to
        // check before trying to display them.
        processView<TextView>(adView, k__price) { view ->
            adView.priceView = view
            val price: CharSequence? = nativeAppInstallAd.price
            if (price != null) {
                view.visibility = View.VISIBLE
                view.text = price
            } else {
                view.visibility = View.INVISIBLE
            }
        }
        processView<RatingBar>(adView, k__star_rating) { view ->
            adView.starRatingView = view
            val starRating = nativeAppInstallAd.starRating
            if (starRating != null) {
                view.rating = starRating.toFloat()
                view.visibility = View.VISIBLE
            } else {
                view.visibility = View.INVISIBLE
            }
        }
        processView<TextView>(adView, k__store) { view ->
            adView.storeView = view
            val store: CharSequence? = nativeAppInstallAd.store
            if (store != null) {
                view.visibility = View.VISIBLE
                view.text = store
            } else {
                view.visibility = View.INVISIBLE
            }
        }

        // Assign native ad object to the native view.
        adView.setNativeAd(nativeAppInstallAd)
    }

    @UiThread
    private fun <T : View> processView(view: View,
                                       key: String,
                                       processor: ViewProcessor<T>): Boolean {
        val id = getIdentifier(key)
        if (id == 0) {
            _logger.error("Can not find identifier for key: $key")
            return false
        }
        val subView = view.findViewById<View>(id)
        if (subView == null) {
            _logger.error("Can not find view for key: $key")
            return false
        }
        processor(subView as T)
        return true
    }

    override fun onAdClosed() {
        _logger.info("${this::onAdClosed}")
        Thread.checkMainThread()
    }

    override fun onAdFailedToLoad(errorCode: Int) {
        _logger.info("${this::onAdFailedToLoad}: code = $errorCode")
        Thread.checkMainThread()
        _bridge.callCpp(_messageHelper.onFailedToLoad, errorCode.toString())
    }

    override fun onAdLeftApplication() {
        _logger.info("${this::onAdLeftApplication}")
        Thread.checkMainThread()
        _bridge.callCpp(_messageHelper.onClicked)
    }

    override fun onAdOpened() {
        _logger.info(this::onAdOpened.name)
        Thread.checkMainThread()
    }

    override fun onAdLoaded() {
        _logger.info("${this::onAdLoaded}")
        Thread.checkMainThread()
        // There is one important difference between the way AdListener objects work with native ads
        // and the way they work with banners and interstitials. Because the AdLoader has its own
        // format-specific listeners (i.e., UnifiedNativeAd.OnUnifiedNativeAdLoadedListener) to use
        // when an ad has loaded, the onAdLoaded() method from AdListener is not called when a
        // native ad loads successfully.
    }

    override fun onAdClicked() {
        _logger.info("${this::onAdClicked}")
        Thread.checkMainThread()
    }

    override fun onAdImpression() {
        _logger.info("${this::onAdImpression}")
        Thread.checkMainThread()
    }
}